## Week 5 Homework 

In this homework we'll put what we learned about Spark in practice.

For this homework we will be using the FHV 2019-10 data found here. [FHV Data](https://github.com/DataTalksClub/nyc-tlc-data/releases/download/fhv/fhv_tripdata_2019-10.csv.gz)

### Question 1: 

**Install Spark and PySpark** 

- Install Spark
- Run PySpark
- Create a local spark session
- Execute spark.version.

What's the output?

#### Answer

In [1]:
import pyspark
from pyspark.sql import SparkSession

spark = SparkSession.builder\
    .master('local[*]') \
    .appName('test')\
    .getOrCreate()

spark.version

Setting default log level to "WARN".
To adjust logging level use sc.setLogLevel(newLevel). For SparkR, use setLogLevel(newLevel).


24/03/02 04:49:17 WARN NativeCodeLoader: Unable to load native-hadoop library for your platform... using builtin-java classes where applicable


'3.3.2'

### Question 2: 

**FHV October 2019**

Read the October 2019 FHV into a Spark Dataframe with a schema as we did in the lessons.

Repartition the Dataframe to 6 partitions and save it to parquet.

What is the average size of the Parquet (ending with .parquet extension) Files that were created (in MB)? Select the answer which most closely matches.

- 1MB
- 6MB
- 25MB
- 87MB

#### Answer

In [2]:
!wget https://github.com/DataTalksClub/nyc-tlc-data/releases/download/fhv/fhv_tripdata_2019-10.csv.gz

--2024-03-02 04:49:20--  https://github.com/DataTalksClub/nyc-tlc-data/releases/download/fhv/fhv_tripdata_2019-10.csv.gz
Resolving github.com (github.com)... 140.82.114.3
Connecting to github.com (github.com)|140.82.114.3|:443... connected.
HTTP request sent, awaiting response... 302 Found
Location: https://objects.githubusercontent.com/github-production-release-asset-2e65be/513814948/efdfcf82-6d5c-44d1-a138-4e8ea3c3a3b6?X-Amz-Algorithm=AWS4-HMAC-SHA256&X-Amz-Credential=AKIAVCODYLSA53PQK4ZA%2F20240302%2Fus-east-1%2Fs3%2Faws4_request&X-Amz-Date=20240302T044834Z&X-Amz-Expires=300&X-Amz-Signature=41a88e0fd03477ff9b6d4736c3f17c5bdee37e5f7617bd8358294289ba261066&X-Amz-SignedHeaders=host&actor_id=0&key_id=0&repo_id=513814948&response-content-disposition=attachment%3B%20filename%3Dfhv_tripdata_2019-10.csv.gz&response-content-type=application%2Foctet-stream [following]
--2024-03-02 04:49:20--  https://objects.githubusercontent.com/github-production-release-asset-2e65be/513814948/efdfcf82-6d5c-

In [3]:
!zcat fhv_tripdata_2019-10.csv.gz | head -n 11 

dispatching_base_num,pickup_datetime,dropOff_datetime,PUlocationID,DOlocationID,SR_Flag,Affiliated_base_number
B00009,2019-10-01 00:23:00,2019-10-01 00:35:00,264,264,,B00009
B00013,2019-10-01 00:11:29,2019-10-01 00:13:22,264,264,,B00013
B00014,2019-10-01 00:11:43,2019-10-01 00:37:20,264,264,,B00014
B00014,2019-10-01 00:56:29,2019-10-01 00:57:47,264,264,,B00014
B00014,2019-10-01 00:23:09,2019-10-01 00:28:27,264,264,,B00014
B00021         ,2019-10-01 00:00:48,2019-10-01 00:07:12,129,129,,B00021         
B00021         ,2019-10-01 00:47:23,2019-10-01 00:53:25,57,57,,B00021         
B00021         ,2019-10-01 00:10:06,2019-10-01 00:19:50,173,173,,B00021         
B00021         ,2019-10-01 00:51:37,2019-10-01 01:06:14,226,226,,B00021         
B00021         ,2019-10-01 00:28:23,2019-10-01 00:34:33,56,56,,B00021         

gzip: stdout: Broken pipe


In [4]:
from pyspark.sql import types

filename = 'fhv_tripdata_2019-10.csv.gz'

schema = types.StructType([
    types.StructField('dispatching_base_num', types.StringType(), True),
    types.StructField('pickup_datetime', types.TimestampType(), True),
    types.StructField('dropOff_datetime', types.TimestampType(), True),
    types.StructField('PUlocationID', types.IntegerType(), True),
    types.StructField('DOlocationID', types.IntegerType(), True),
    types.StructField('SR_Flag', types.StringType(), True),
    types.StructField('Affiliated_base_number', types.StringType(), True),
])

df_fhv = spark.read \
    .option("header", "true") \
    .schema(schema) \
    .csv(filename)

df_fhv = df_fhv.repartition(6)

df_fhv.write.parquet('fhv/2019/10/')

In [5]:
!ls -ltrh fhv/2019/10/*parquet

-rw-r--r-- 1 minasonbol minasonbol 6.4M Mar  2 04:49 fhv/2019/10/part-00003-b4a20026-8748-4b64-9432-5cb3cd4ca146-c000.snappy.parquet
-rw-r--r-- 1 minasonbol minasonbol 6.4M Mar  2 04:49 fhv/2019/10/part-00001-b4a20026-8748-4b64-9432-5cb3cd4ca146-c000.snappy.parquet
-rw-r--r-- 1 minasonbol minasonbol 6.4M Mar  2 04:49 fhv/2019/10/part-00000-b4a20026-8748-4b64-9432-5cb3cd4ca146-c000.snappy.parquet
-rw-r--r-- 1 minasonbol minasonbol 6.4M Mar  2 04:49 fhv/2019/10/part-00002-b4a20026-8748-4b64-9432-5cb3cd4ca146-c000.snappy.parquet
-rw-r--r-- 1 minasonbol minasonbol 6.4M Mar  2 04:49 fhv/2019/10/part-00004-b4a20026-8748-4b64-9432-5cb3cd4ca146-c000.snappy.parquet
-rw-r--r-- 1 minasonbol minasonbol 6.4M Mar  2 04:49 fhv/2019/10/part-00005-b4a20026-8748-4b64-9432-5cb3cd4ca146-c000.snappy.parquet


### Question 3: 

**Count records** 

How many taxi trips were there on the 15th of October?

Consider only trips that started on the 15th of October.

- 108,164
- 12,856
- 452,470
- 62,610

> [!IMPORTANT]
> Be aware of columns order when defining schema

In [6]:
from pyspark.sql import functions as F

# convert pickup_datetime column to a Date type
df_fhv = df_fhv.withColumn('pickup_date', F.to_date(df_fhv.pickup_datetime))

# filter on the 2019-10-15 + count
df_fhv.filter(df_fhv.pickup_date == "2019-10-15").count()

62610

### Question 4: 

**Longest trip for each day** 

What is the length of the longest trip in the dataset in hours?

- 631,152.50 Hours
- 243.44 Hours
- 7.68 Hours
- 3.32 Hours


#### Answer

In [7]:
timeFmt = "yyyy-MM-dd HH24:mm:ss"

timeDiff = (F.unix_timestamp('dropOff_datetime', format=timeFmt)
            - F.unix_timestamp('pickup_datetime', format=timeFmt))

df_fhv.withColumn("duration", timeDiff).agg({"duration": "max"}).collect()[0][0] / 3600

631152.5

### Question 5: 

**User Interface**

Spark’s User Interface which shows the application's dashboard runs on which local port?

- 80
- 443
- 4040
- 8080

#### Answer

4040

### Question 6: 

**Least frequent pickup location zone**

Load the zone lookup data into a temp view in Spark</br>
[Zone Data](https://github.com/DataTalksClub/nyc-tlc-data/releases/download/misc/taxi_zone_lookup.csv)

Using the zone lookup data and the FHV October 2019 data, what is the name of the LEAST frequent pickup location Zone?</br>

- East Chelsea
- Jamaica Bay
- Union Sq
- Crown Heights North

In [8]:
!wget https://github.com/DataTalksClub/nyc-tlc-data/releases/download/misc/taxi_zone_lookup.csv

--2024-03-02 04:50:19--  https://github.com/DataTalksClub/nyc-tlc-data/releases/download/misc/taxi_zone_lookup.csv
Resolving github.com (github.com)... 140.82.114.4
Connecting to github.com (github.com)|140.82.114.4|:443... connected.
HTTP request sent, awaiting response... 302 Found
Location: https://objects.githubusercontent.com/github-production-release-asset-2e65be/513814948/5a2cc2f5-b4cd-4584-9c62-a6ea97ed0e6a?X-Amz-Algorithm=AWS4-HMAC-SHA256&X-Amz-Credential=AKIAVCODYLSA53PQK4ZA%2F20240302%2Fus-east-1%2Fs3%2Faws4_request&X-Amz-Date=20240302T045019Z&X-Amz-Expires=300&X-Amz-Signature=799871f2e6e4b2036343aa776ec5a4ce33d47705e6ac7972e2acf60e75db5ac6&X-Amz-SignedHeaders=host&actor_id=0&key_id=0&repo_id=513814948&response-content-disposition=attachment%3B%20filename%3Dtaxi_zone_lookup.csv&response-content-type=application%2Foctet-stream [following]
--2024-03-02 04:50:19--  https://objects.githubusercontent.com/github-production-release-asset-2e65be/513814948/5a2cc2f5-b4cd-4584-9c62-a6e

In [14]:
!head taxi_zone_lookup.csv

In [9]:
df_zones = spark.read \
    .option("header", "true") \
    .csv('taxi_zone_lookup.csv')

df_zones.show()

+----------+-------------+--------------------+------------+
|LocationID|      Borough|                Zone|service_zone|
+----------+-------------+--------------------+------------+
|         1|          EWR|      Newark Airport|         EWR|
|         2|       Queens|         Jamaica Bay|   Boro Zone|
|         3|        Bronx|Allerton/Pelham G...|   Boro Zone|
|         4|    Manhattan|       Alphabet City| Yellow Zone|
|         5|Staten Island|       Arden Heights|   Boro Zone|
|         6|Staten Island|Arrochar/Fort Wad...|   Boro Zone|
|         7|       Queens|             Astoria|   Boro Zone|
|         8|       Queens|        Astoria Park|   Boro Zone|
|         9|       Queens|          Auburndale|   Boro Zone|
|        10|       Queens|        Baisley Park|   Boro Zone|
|        11|     Brooklyn|          Bath Beach|   Boro Zone|
|        12|    Manhattan|        Battery Park| Yellow Zone|
|        13|    Manhattan|   Battery Park City| Yellow Zone|
|        14|     Brookly

In [10]:
df_zones.write.parquet('zones')

In [12]:
df_zones = spark.read.parquet('zones/')
df_result = df_fhv.join(df_zones, df_fhv.PUlocationID == df_zones.LocationID)

In [13]:
# Counting occurrences of each zone
zone_counts = df_result.groupBy("Zone").count()

In [18]:
# Finding the least frequent pickup location zone
zone_counts.orderBy("count").show()

+--------------------+-----+
|                Zone|count|
+--------------------+-----+
|         Jamaica Bay|    1|
|Governor's Island...|    2|
| Green-Wood Cemetery|    5|
|       Broad Channel|    8|
|     Highbridge Park|   14|
|        Battery Park|   15|
|Saint Michaels Ce...|   23|
|Breezy Point/Fort...|   25|
|Marine Park/Floyd...|   26|
|        Astoria Park|   29|
|    Inwood Hill Park|   39|
|       Willets Point|   47|
|Forest Park/Highl...|   53|
|  Brooklyn Navy Yard|   57|
|        Crotona Park|   62|
|        Country Club|   77|
|     Freshkills Park|   89|
|       Prospect Park|   98|
|     Columbia Street|  105|
|  South Williamsburg|  110|
+--------------------+-----+
only showing top 20 rows

